airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py

```diff
@@ -0,0 +1,1387 @@
+                                            __typename
+                                            usageCount: asyncUsageCount
+                                            code
+                                            id
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    """
+
+    query_name = "codeDiscountNodes"
+    sort_key = "UPDATED_AT"
+
+    code_discount_fields: List[Field] = [
+        Field(name="discountClass", alias="discountType"),
+        Field(
+            name="codes",
+            fields=[
+                Field(
+                    name="edges",
+                    fields=[
+                        Field(
+                            name="node",
+                            fields=[
+                                "__typename",
+                                Field(name="asyncUsageCount", alias="usageCount"),
+                                "code",
+                                "id",
+                            ],
+                        )
+                    ],
+                )
+            ],
+        ),
+    ]
+
+    code_discount_fragments: List[InlineFragment] = [
+        # the type: DiscountCodeApp has no `"summary"` field available
+        InlineFragment(type="DiscountCodeApp", fields=["updatedAt", "createdAt", *code_discount_fields]),
+        InlineFragment(type="DiscountCodeBasic", fields=["updatedAt", "createdAt", "summary", *code_discount_fields]),
+        InlineFragment(type="DiscountCodeBxgy", fields=["updatedAt", "createdAt", "summary", *code_discount_fields]),
+        InlineFragment(type="DiscountCodeFreeShipping", fields=["updatedAt", "createdAt", "summary", *code_discount_fields]),
+    ]
+
+    query_nodes: List[Field] = [
+        "__typename",
+        "id",
+        Field(name="codeDiscount", fields=code_discount_fragments),
+    ]
+
+    record_composition = {
+        "new_record": "DiscountCodeNode",
+        # each DiscountCodeNode has `DiscountRedeemCode`
+        "record_components": ["DiscountRedeemCode"],
+    }
+
+    def record_process_components(self, record: Optional[Mapping[str, Any]] = None) -> Optional[Iterable[Mapping[str, Any]]]:
+        """
+        Defiines how to process collected components.
+        """
+        if record:
+            record_components = record.get("record_components", {})
+            if record_components:
+                discounts = record_components.get("DiscountRedeemCode", [])
+                if len(discounts) > 0:
+                    for discount in discounts:
+                        # resolve parent id from `str` to `int`
+                        discount["admin_graphql_api_id"] = discount.get("id")
+                        discount["price_rule_id"] = self.tools.resolve_str_id(discount.get(BULK_PARENT_KEY))
```

Comment:
do we expect discount.get(BULK_PARENT_KEY) to ever be null?

Reply:
No.